File: Assignment 1/q2(c).cpp
// q2(c).cpp

#include <iostream>
#include <vector>
#include <chrono>

using namespace std;
using namespace chrono;

//matrix product function 
void multiplyMatrices(vector<vector<int>>& A,vector<vector<int>>&B, vector<vector<int>>&C, int N) {
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            C[i][j] = 0;
            for (int k = 0; k < N; ++k) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
}

int main() {
    int N = 1024; //{64, 128, 256, 512, 1024};

    vector<vector<int>> A(N, vector<int>(N, 1.0));
    vector<vector<int>> B(N, vector<int>(N, 1.0));
    vector<vector<int>> C(N, vector<int>(N, 0.0));

    auto start = high_resolution_clock::now();
    multiplyMatrices(A, B, C, N);
    auto end = high_resolution_clock::now();

    auto duration = duration_cast<milliseconds>(end - start);

    cout << "Matrix size: " << N << "x" << N << ", Meat Time taken: " << duration.count() << " ms" << endl;
    return 0;
}

File: Assignment 1/q2(c).py
import numpy as np
import time

def multiply_matrices(A, B, N):
    C = np.zeros((N, N), dtype=int)
    for i in range(N):
        for j in range(N):
            for k in range(N):
                C[i][j] += A[i][k] * B[k][j]
    return C

N = 1024#[64, 128, 256, 512, 1024]
A = np.ones((N, N), dtype=int)
B = np.ones((N, N), dtype=int)

start_time = time.time()
C = multiply_matrices(A, B, N)
end_time = time.time()

duration = (end_time - start_time) * 1000  

print(f"Matrix size: {N}x{N}, Time taken: {duration:.2f} ms")